examples/cfq/models.py

```diff
@@ -212,32 +217,33 @@ def apply(self,
         dropout_rate=vertical_dropout_rate)
 
     def decode_step_fn(carry, x):
-      rng, multilayer_lstm_output, last_prediction = carry
-      previous_states, h = multilayer_lstm_output
+      rng, multilayer_lstm_output, last_prediction, prev_attention = carry
+      previous_states, _ = multilayer_lstm_output
       carry_rng, categorical_rng = jax.random.split(rng, 2)
       if not train:
         x = last_prediction
       x = shared_embedding(x)
-      dec_prev_state = jnp.expand_dims(h, 1)
-      attention = mlp_attention(dec_prev_state, projected_keys,
-                                encoder_hidden_states, attention_mask)
-      lstm_input = jnp.concatenate([x, attention], axis=-1)
+      lstm_input = jnp.concatenate([x, prev_attention], axis=-1)
       states, h = multilayer_lstm_cell(horizontal_dropout_masks=h_dropout_masks,
                                        vertical_dropout_masks=v_dropout_masks,
                                        input=lstm_input,
                                        previous_states=previous_states)
-      inner_proj_input = jnp.concatenate([x, attention, h], axis=-1)
-      pre_output = pre_output_layer(inner_proj_input)
-      logits = projection(pre_output)
+      context = mlp_attention(jnp.expand_dims(h, 1), projected_keys,
+                              encoder_hidden_states, attention_mask)
+      context_and_state = jnp.concatenate([context, h], axis=-1)
+      attention = jnp.tanh(attention_layer(context_and_state))
+      logits = projection(attention)
       predicted_tokens = jax.random.categorical(categorical_rng, logits)
       predicted_tokens_uint8 = jnp.asarray(predicted_tokens, dtype=jnp.uint8)
-      return (carry_rng, (states, h),
-              predicted_tokens_uint8), (logits, predicted_tokens_uint8)
+      new_carry = (carry_rng, (states, h), predicted_tokens_uint8, attention)
```

Comment:
I guess we don't have to pass `h` here, since we don't use it in the next step anyway.